code/modules/cm_aliens/hivebuffs/hivebuff.dm

```diff
@@ -0,0 +1,517 @@
+		return FALSE
+
+	// _seek_queen_approval() includes a 20 second timeout so we check that everything still exists that we need.
+	if(QDELETED(purchased_pylon) || QDELETED(purchasing_mob) && !purchasing_mob.check_state())
+		return FALSE
+
+	// Actually process the buff and apply effects - If the buff succeeds engage_message will return TRUE, if it fails there should be an engage_failure_message set.
+	if(!on_engage(purchased_pylon))
+		if(engage_failure_message && istext(engage_failure_message))
+			to_chat(purchasing_mob, SPAN_XENONOTICE(engage_failure_message))
+			to_chat(hive.living_xeno_queen, SPAN_XENONOTICE(engage_failure_message))
+			return
+		else
+			stack_trace("[purchasing_mob] attempted to purchase a hive buff: [name] and failed to engage and returned an invalid failure message or no failure message.")
+			return
+
+	// All checks have passed.
+
+	// Purchase and deduct funds only after we're sure the buff has engaged
+	if(!_purchase_and_deduct(purchasing_mob))
+		return
+
+	for(var/mob/living/carbon/xenomorph/xeno in hive.totalXenos)
+		if(apply_on_new_xeno)
+			RegisterSignal(SSdcs, COMSIG_GLOB_XENO_SPAWN, PROC_REF(_handle_xenomorph_new))
+		apply_buff_effects(xeno)
+
+
+	log_admin("[purchasing_mob] and [hive.living_xeno_queen] of [hive.hivenumber] have purchased a hive buff: [name].")
+
+	// Add to the relevant hive lists.
+	LAZYADD(hive.used_hivebuffs, src)
+	LAZYADD(hive.active_hivebuffs, src)
+
+	// Register signal to check if the pylon is ever destroyed.
+
+	// for(var/obj/effect/alien/resin/special/pylon/endgame/pylon_to_register in pylons_to_use)
+	// 	LAZYADD(sustained_pylons, purchased_pylon)
+	// 	pylon_to_register.sustain_hivebuff(src)
+	// 	RegisterSignal(pylon_to_register, COMSIG_PARENT_QDELETING, PROC_REF(_on_pylon_deletion))
+
+	// Announce to our hive that we've completed.
+	_announce_buff_engage()
+
+	// If we need a timer to call _on_cease() we add it here and store the id, used for deleting the timer if we Destroy().
+	// If we have no duration to the buff then we call _on_cease() immediately.
+	if(duration)
+		_timer_id = addtimer(CALLBACK(src, PROC_REF(_on_cease)), duration, TIMER_STOPPABLE)
+	else
+		_on_cease()
+	return TRUE
+
+/// Behaviour for the buff goes in here.
+/// IMPORTANT: If you buff has any kind of conditions which can fail. Set an engage_failure_message and return FALSE.
+/// If your buff succeeds you must return TRUE
+/datum/hivebuff/proc/on_engage(obj/effect/alien/resin/special/pylon/purchased_pylon)
+	return TRUE
+
+/// Wrapper for on_cease(), calls qdel(src) after on_cease() behaviour.
+/datum/hivebuff/proc/_on_cease()
+	_announce_buff_cease()
+	/// Clear refernces to this buff and unregister signal
+	on_cease()
+	if(!ended_via_pylon_qdeletion)
+		for(var/obj/effect/alien/resin/special/pylon/endgame/pylon_to_clear in sustained_pylons)
+			pylon_to_clear.remove_hivebuff()
+			UnregisterSignal(pylon_to_clear, COMSIG_PARENT_QDELETING)
```

Comment:
Currently this is unnecessary to register because the registration is commented out.